### DISCUSSION TITLE
Piercing

### DISCUSSION THREAD

**Discussion by BenLubar** [ReactiveDrop/reactivedrop_public_src#310] [Ideas]
In Alien Swarm, hitscan weapons (mostly weapons that fire bullets) have a chance to pierce as well as a maximum number of times they can pierce per shot. This is confusing, and to make it worse, some weapons interact with the Piercing skill and some do not.

In the following list, skill% means for 40% for Wildcat, 60% for Wolfe, and 0% for other marines.

**22A3-1 Assault Rifle:** skill%, max 1
**22A7-Z Prototype Assault Rifle:** skill%, max 1
**S23A SynTek Autogun:** skill%, max 1
**M42 Vindicator:** skill%, max 1
**M73 Twin Pistols:** skill%, max 1
**IAF Advanced Sentry Gun:** cannot pierce
**Model 35 Pump-action Shotgun:** 100%, max 0
**Precision Rail Rifle:** 100%, max 10
**K80 Personal Defense Weapon:** skill%, max 1
**IAF Minigun:** skill%, add 20% if firing for more than 3 seconds, max 1
**AVK-36 Marksman Rifle:** 350%, max 1/2/3 for normal/zoomed/damage-amped
**PS50 Bulldog:** skill%, max 1
**IAF HAS42 Devastator:** skill%, max 1 on special weapons; 100%, max 0 on other marines
**22A4-2 Combat Rifle:** skill%, max 1
**22A4-2 Combat Rifle (alt fire):** 100%, max 0
**22A5 Heavy Assault Rifle:** skill%, max 1
**IAF Medical SMG:** skill%, max 1

Bullets can pierce aliens, including dead frozen aliens, but not including alive shieldbugs.

They can also pierce glass. *Any bullet that pierces glass changes to a skill%, max 1 piercing bullet.* Glass can only be pierced if it is breakable and at most 16 units thick.

Bullets with over 100% piercing chance can penetrate *anything* (even things that were excluded above, except for breakable glass), losing 100% piercing chance each time. If the bullet isn't hitting a static part of the world here, it has a maximum of 100% chance to pierce going forward. This includes when it hits an alien.

If the thing the bullet hit is within 70 units of the marine vertically, the bullet leaves going parallel to the ground.

If a bullet hits a static part of the world, it does *not* count towards the maximum number of times a bullet can pierce and the bullet teleports forward 28 units.

Example of non-intuitive behavior:

> A damage-amped AVK-36 Marksman Rifle has several drones in front of it.
>
> Before hitting the first drone: 350% chance to pierce, 3 max.
> After first drone: 100% chance to pierce, 2 max.
> Second drone: 100% chance to pierce, 1 max.
> Third drone: 100% chance to pierce, 0 max.
> Fourth drone: the bullet hits the drone and stops.
>
> Now, the same weapon with the same damage amp fires again, this time aiming through a very thick (64 unit) wall.
>
> Before hitting the wall: 350% chance to pierce, 3 max.
> After hitting the wall: bullet teleports 28 units into the wall, piercing 250%, 3 max.
> Continuing: the bullet is now 56 units into the wall, piercing 150%, 3 max.
> After that: the bullet is now 20 units past the wall, piercing 50%, 3 max.
> The process continues the same as above, as if the bullet had just been fired, but with a 50% chance to stop after each drone.

In short, this system is a mess, and very complicated to explain.

So let's look for some alternatives.

I've seen two systems used in other games:

System A: Attacks can either pierce or they can't. There is possibly a limit to how many targets a single attack can hit. For example, the sniper rifle could penetrate up to 5 targets. The shotgun might penetrate 1 target. Note that this does not handle piercing through walls.

System B: Attacks have a limit of how much material they can pierce through, with the possibility of ignoring the limit when hitting an enemy. For example, the railgun might be able to penetrate 48 units of material, but keep going if it hits an enemy. If the flamer was hitscan, it would penetrate 0 units of material and ignore enemies. The minigun might be able to pierce 96 units of material (a drone's hitbox is about 53 units thick from the front, so this could penetrate a drone, a wall, and then hit another alien on the other side).

The third alternative is to keep the current system, but simplify its logic.

This topic has also been posted on the [Steam discussion forum](https://steamcommunity.com/app/563560/discussions/2/3361397532260314955/).

**Comment by ywgATustcbbs** [comment]
My idea is, Each bullet has a basic momentum, different materials have different resistance per unit thickness. Skills or amp will increase the momentum by a random percentage from 0-skill%. While walls, drones will decrease the momentum by resistance times thickness. If the momentum is less than 0, it stops and no damage is done. 

For example,
1. A bullet has a basic 100 momentum (varies on different weapons)
2. Skill and amp increased the "chance" to 350%(+250%)
3. Take a random number between 0-250%, Such as 150%
4. Adjusted momentum is 100 basic +150 additional =250
5. Drone resistance is 1.5 per unit thickness, typical hitbox is 50, It's total resistance is 1.5*50=75
6. Bullet hits 1st Drone, momentum >0, deals damage. Then the momentumchanges from 250 to 175(decreases 75) >0. Thus it penetrates the Drone.
7. Repeat step 6 for 2nd drone. Deals damage and penetrates. 175->100
8. Same for the 3rd drone,. 100->25
9. 4Th done, only deals damage. 25->-50
10. Similar for a wall. But resistance could be set to another value.

Advantages:
1. Emulates the reality
2. Only one rand() per bullet, other values could be pre-calculated
3. Rand() is linear distribution by default, But could be replaced by other distribution functions to get a more realistic felling. Such ad normal distribution, exponential distribution. This could be optimized for speed by using a linear to non-linear lookup table
4. If the actual piercing distance is provided for mobs, It can emulate side piercing effect that a bullet deals less damage (compared to full hitbox penetration) but has more (possible) piercing numbers. It is reasonable that shooting alien's feet/claws will deal less damage but has small resistance on bullet movement.
5. If needed, It can also emulate bullet damage decreasing with momentum (linear or non-linear)

**Comment by squillisky** [comment]
I think that the original system is good regarding the number and probability to penetrate drones. But for walls I suggest to sum number of penetrable drones multiply the converted number of unit per drone (53 units) and then half the result. For example railgun can penetrate 10 drones so (10*53)/2= 265 unit of wall. But if it penetrate 1 drome the maximum amount of penetrable wall become (9*53)/2= 238.5 unit. If you want you could change the denominator for different materials. I think it would be an easy system but effective.
